src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -625,32 +642,32 @@ void ObjectCopyRequest<I>::merge_write_ops() {
 template <typename I>
 void ObjectCopyRequest<I>::compute_zero_ops() {
   ldout(m_cct, 20) << dendl;
-
   m_src_image_ctx->image_lock.lock_shared();
   bool hide_parent = (m_src_snap_id_start == 0 &&
                       m_src_image_ctx->parent != nullptr);
   m_src_image_ctx->image_lock.unlock_shared();
-
+  bool update_map_nonexistent = false;
   // ensure we have a zeroed interval for each snapshot
   for (auto& [src_snap_seq, _] : m_snap_map) {
     if (m_src_snap_id_start < src_snap_seq) {
       m_dst_zero_interval[src_snap_seq];
     }
   }
-
   // exists if copying from an arbitrary snapshot w/o any deltas in the
   // start snapshot slot (i.e. DNE)
   bool object_exists = (
       m_src_snap_id_start > 0 &&
       m_snapshot_delta.count({m_src_snap_id_start, m_src_snap_id_start}) == 0);
-  bool fast_diff = m_dst_image_ctx->test_features(RBD_FEATURE_FAST_DIFF);
-  uint64_t prev_end_size = 0;
 
+  bool fast_diff = m_dst_image_ctx->test_features(RBD_FEATURE_FAST_DIFF);
+  uint64_t prev_end_size = compute_starting_end_size();
```

Comment:
@idryomov I am not sure I understand your example, so I will just try to explain the logic (as I currently understand it). So if it is "zero interval inside the object" case, we just insert this extent, I think there are no questions about this? If it is "zero interval at the object end" case, then we may ignore this extent (no need to write zeros at the object end) unless the previous snapshot end is larger than the start of this zero extent: in this case we need to insert zero extent to hide data on the previous snapshot.

Does it make sense to you? Do you think it does not work as I described?

Reply:
IIUC, in your case, if the zeroed extent begins at 2M and end size = 3M, there is no requirement to insert it as the previous object size was 1M.

Reply:
> If it is "zero interval at the object end" case, then we may ignore this extent (no need to write zeros at the object end)

Conceptually, I get that writing zeroes at the object end is unnecessary, but the condition here isn't `object_extent.offset == end_size` but rather `object_extent.offset + object_extent.length >= end_size`.  Perhaps the bit that I'm missing is the details of the scenario when `end_size` ends up being larger than `prev_end_size`?  Here is the code in question:
```
    // update end_size if there are writes into higher offsets
    uint64_t end_size = prev_end_size;
    auto iter = m_snapshot_sparse_bufferlist.find(src_snap_seq);
    if (iter != m_snapshot_sparse_bufferlist.end()) {
      for (auto &sparse_bufferlist : iter->second) {
        object_exists = true;
        end_size = std::max(
          end_size, sparse_bufferlist.get_off() + sparse_bufferlist.get_len());
      }
    }
```
Going back to my example for illustration: if `prev_end_size` is 1M, `end_size` is initialized to 1M but gets increased to 3M in this loop, is it possible to encounter e.g. 2M~1M (i.e. ending at new `end_size`) zero extent?  If so, why is it OK to ignore it?

> unless the previous snapshot end is larger than the start of this zero extent: in this case we need to insert zero extent to hide data on the previous snapshot.

... but we are inserting the entire zero extent, including the "unnecessary" portion that starts at `prev_end_size` -- the portion that would have been ignored had the extent started at `prev_end_size`.

The other thing that I'm missing is `end_size` getting decreased even if the zero extent is ignored:
```
          end_size = std::min(end_size, object_extent.offset);
```

> 
> Does it make sense to you? Do you think it does not work as I described?

So far this code doesn't line up in my head.  It could be that I have the wrong mental model around it, hence all these questions ;)